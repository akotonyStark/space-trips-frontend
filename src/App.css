html,
body {
  height: 100%;
  width: 100%;
  margin: 0;
}

html {
  scroll-behavior: smooth;
}

.App {
  width: 70%;
  margin: 0 auto;
  border: 1px solid #e3e3e3;
}

.container {
  display: flex;
  flex-direction: row;
}

.space-header {
  display: flex;
  align-items: center;
  /* justify-content: center; */
  width: 35%;
  height: 88px;
  background: #fff;
  font-size: 24px;
  font-family: monospace;
}

.space-body {
  width: 35%;
  height: 85vh;
  background: #fafafb;
  max-height: 85vh;
  overflow: auto;
}
.map-header {
  display: flex;
  align-items: center;
  width: 65%;
  height: 88px;
  background: #333333;
  color: white;
}

.map-body {
  width: 65%;
  height: 95vh;
  background: #000;
}

.trip-card:hover {
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
  cursor: pointer;
}

.trip-title {
  font-size: 16px;
  font-weight: bold;
}

.trip-planet {
  font-size: 12px;
}

.mapboxgl-popup-close-button {
  display: none !important;
}

.mapboxgl-popup-content {
  padding: 0px !important;
}

.react-datetime-picker__inputGroup__input {
  color: white;
  border: none;
}

.datetime-picker {
  border: 0px;
  background-color: inherit;
}

@media screen and (max-width: 1600px) {
  .App {
    width: 100%;
    margin: 0 auto;
  }
}
