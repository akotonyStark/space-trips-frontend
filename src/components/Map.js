import React, { useState, useMemo, useContext, useEffect, useRef } from "react";
import Map, {
  Marker,
  Popup,
  NavigationControl,
  FullscreenControl,
  ScaleControl,
  GeolocateControl,
} from "react-map-gl";

import YellowPin from "../assets/marker.png";
import RedPin from "../assets/red.png";
import space_image from "../assets/space_image.png";

import { AppContext } from "../App";
import { convertToID } from "../utils/helperFunctions";

const MapBody = () => {
  const [popupInfo, setPopupInfo] = useState(null);
  const { trips, viewState, setViewState, hovered, setMarker, mapCenter } =
    useContext(AppContext);

  const markerRef = useRef(null);
  const mapRef = useRef(null);

  useEffect(() => {
    // Using easeTo options.
    if (mapRef.current) {
      mapRef.current.easeTo({
        center: mapCenter,
        zoom: 3.5,
        speed: 0.3,
        curve: 1,
        duration: 5000,
        easing(t) {
          return t;
        },
      });
    }
  }, [mapCenter]);

  const handleMarkerInteraction = (sp_center) => {
    setPopupInfo({ ...sp_center, image: space_image });
    setMarker({
      id: convertToID(sp_center.name),
      isBouncing: true,
    });

    let elementId = convertToID(sp_center.name);
    let element = document.getElementById(`${elementId}`);

    // scroll to element
    element.scrollIntoView({
      behavior: "smooth",
    });

    // element.style.border = "1px solid gold";
    element.style.boxShadow = "0 4px 8px 0 rgba(0, 0, 0, 0.2)";

    //stop bouncing after 3 seconds
    setTimeout(() => {
      setMarker({ id: convertToID(sp_center.name), isBouncing: false });
      element.style.boxShadow = null;
    }, 3000);
  };

  const pins = useMemo(
    () =>
      trips.map((center, index) => (
        <Marker
          key={`marker-${index}`}
          longitude={center.longitude}
          latitude={center.latitude}
          anchor="bottom"
        >
          {hovered.state && hovered.id === convertToID(center.name) ? (
            <img
              ref={markerRef}
              src={RedPin}
              alt="red-marker"
              onClick={() => handleMarkerInteraction(center)}
            />
          ) : (
            <img
              ref={markerRef}
              src={YellowPin}
              alt="yellow-marker"
              onClick={() => handleMarkerInteraction(center)}
            />
          )}
        </Marker>
      )),
    [trips, hovered]
  );

  return (
    <div className="map-body">
      <Map
        ref={mapRef}
        name="space-centers-map"
        initialViewState={viewState}
        onMove={(evt) => setViewState(evt.viewState)}
        mapStyle="mapbox://styles/mapbox/dark-v8"
        mapboxAccessToken={process.env.REACT_APP_MAPBOX_ACCESS_TOKEN}
      >
        <GeolocateControl position="top-right" />
        <NavigationControl position="top-right" />
        <FullscreenControl position="top-right" />
        <ScaleControl />

        {popupInfo && (
          <Popup
            style={{ cursor: "pointer" }}
            anchor="bottom"
            longitude={Number(popupInfo.longitude)}
            latitude={Number(popupInfo.latitude)}
            closeOnClick={false}
            onClose={() => setPopupInfo(null)}
          >
            <img
              width="100%"
              src={popupInfo.image}
              alt="marker"
              style={{ height: 150 }}
              onClick={() => setPopupInfo(null)}
              title={popupInfo.description}
            />
            <div
              style={{
                height: 30,
                fontFamily: "Lato",
                textAlign: "center",
                fontWeight: "bold",
              }}
            >
              {popupInfo.name} -{" "}
              <span style={{ color: "gold" }}>{popupInfo.planet.name}</span>
            </div>
          </Popup>
        )}
        {pins}
      </Map>
    </div>
  );
};

export default MapBody;
