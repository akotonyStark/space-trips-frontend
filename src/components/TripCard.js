import React, { useContext } from "react";
import styled from "styled-components";
import rocket from "../assets/icons/Rocket.svg";
import { bounce } from "react-animations";
import { keyframes } from "styled-components";

import { gql } from "@apollo/client";
import { AppContext } from "../App";
import { FLIGHTS } from "../data/store";
import { convertToID, client } from "../utils/helperFunctions";

const Bounce = styled.img`
  animation: 1s ${keyframes`${bounce}`} infinite;
`;

const StyledCard = styled.div`
  width: 80%;
  margin: 0 auto;
  margin-top: 35px;
  height: 200px;
  background-color: #fff;
  border: "1px solid black";
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  font-family: "Lato";
  transition: transform 0.5s;
`;

const StyledButton = styled.div`
  width: 100%;
  height: 48px;
  background-color: black;
  color: white;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  font-size: 12px;
`;

const TripCard = ({ spaceCenter }) => {
  const {
    setHovered,
    marker,
    setMapCenter,
    departureDate,
    setFlights,
    setShowFlightsList,
    setshowSearching,
    showNotification,
    setResponse,
  } = useContext(AppContext);

  const handleGetFlights = (fromID) => {
    setshowSearching(true);
    setShowFlightsList(true);

    let departure = new Date(departureDate).toISOString().slice(0, 10);
    client
      .query({
        query: gql`
          # Write your query or mutation here
          query GetFlightsList {
            flights(from: ${fromID}, departureDay: "${departure}", pageSize: 100) {
              nodes {
                id
                code
                launchSite {
                  id
                  uid
                  name
                  description
                  latitude
                  longitude
                  planet {
                    id
                    name
                  }
                }
                landingSite {
                  id
                  uid
                  name
                  description
                  latitude
                  longitude
                  planet {
                    id
                    name
                  }
                }
                departureAt
                seatCount
                availableSeats
              }
            }
          }
        `,
      })
      .then((res) => {
        setFlights(res.data.flights.nodes);

        if (res.data.flights.nodes.length === 0) {
          setshowSearching(null);
          showNotification(true);
          setResponse(
            "No flights available at this space station on this date."
          );
        } else {
          setshowSearching(false);
        }
      })
      .catch((error) => {
        setFlights(FLIGHTS.nodes);
        if (FLIGHTS.nodes.length === 0) {
          setshowSearching(null);
          showNotification(true);
          setResponse(
            "No flights available at this space station on this date."
          );
        } else {
          setshowSearching(false);
        }
      });
  };

  return (
    <div>
      <StyledCard
        className="trip-card"
        id={convertToID(spaceCenter.name)}
        onMouseEnter={() =>
          setHovered({
            id: convertToID(spaceCenter.name),
            state: true,
          })
        }
        onMouseLeave={() =>
          setHovered({
            id: convertToID(spaceCenter.name),
            state: false,
          })
        }
        onClick={() =>
          setMapCenter([spaceCenter.longitude, spaceCenter.latitude])
        }
      >
        <div style={{ padding: 10 }}>
          <div
            style={{
              display: "flex",
              justifyContent: "space-between",
              alignItems: "center",
            }}
          >
            <span className="trip-title">{spaceCenter.name}</span>
            {marker.isBouncing &&
            marker.id === convertToID(spaceCenter.name) ? (
              <Bounce src={rocket} alt={rocket} className="rocket" />
            ) : (
              <img src={rocket} alt={rocket} className="rocket" />
            )}
          </div>
          <span className="trip-planet">{spaceCenter.planet.name}</span>
          <p style={{ marginTop: 30 }}>12 departures planned today</p>
        </div>

        <StyledButton
          title="click to view all flights"
          onClick={() => handleGetFlights(spaceCenter.id)}
        >
          SEE ALL FLIGHTS
        </StyledButton>
      </StyledCard>
    </div>
  );
};

export default TripCard;
