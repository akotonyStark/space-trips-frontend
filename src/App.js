import "./App.css";
import Map from "./components/Map";
import SearchBar from "./components/SearchBar";
import TripsList from "./components/List";
import TripsHeader from "./components/TripsHeader";
import React, { createContext, useState } from "react";
import { ApolloClient, InMemoryCache, gql, HttpLink } from "@apollo/client";

import { TRIPS } from "./data/store.js";
import Modal from "./components/Modal";

export const AppContext = createContext();

const link = new HttpLink({
  uri: process.env.REACT_APP_API_URL,
  credentials: "same-origin",
  headers: {
    authorization: "Bearer API_KEY",
  },
});
const client = new ApolloClient({
  link: link,
  cache: new InMemoryCache(),
});

const INIT_STATE = {
  width: "100%",
  height: 900,
  latitude: 41.579606918652054,
  longitude: 4.244298260567439,
  zoom: 3.5,
  bearing: 0,
  pitch: 0,
  transitionDuration: 1000,
};

function App() {
  const [page, setPage] = React.useState(1);
  const [spaceCenters, setSpaceCenters] = React.useState([]);
  const [trips, setTrips] = React.useState([]);
  const [flights, setFlights] = React.useState([]);
  const [viewState, setViewState] = React.useState(INIT_STATE);
  const [hovered, setHovered] = useState({ id: "", state: false });
  const [marker, setMarker] = React.useState({ id: "", isBouncing: false });
  const [mapCenter, setMapCenter] = React.useState([
    INIT_STATE.longitude,
    INIT_STATE.latitude,
  ]);
  const [isModalOpen, setIsModalOpen] = React.useState(false);

  const getSpaceTrips = (page) => {
    client
      .query({
        query: gql`
          query GetSpaceCenters {
            spaceCenters(page: ${page}, pageSize: 100) {
              nodes {
                id
                uid
                name
                description
                latitude
                longitude
                planet {
                  id
                  name
                  code
                }
              }
              pagination {
                total
                page
                pageSize
              }
            }
          }
        `,
      })
      .then((result) => {
        setTrips(result.data.spaceCenters.nodes);
      })
      .catch((err) => {
        console.log("could not load data");
        console.log("...reverting to offline data");
        setTrips(TRIPS.spaceCenters.nodes);
      });
  };

  const getAllFlights = () => {};

  React.useEffect(() => {
    getSpaceTrips(page);
  }, [page]);

  return (
    <AppContext.Provider
      value={[
        spaceCenters,
        setSpaceCenters,
        trips,
        setTrips,
        viewState,
        setViewState,
        hovered,
        setHovered,
        marker,
        setMarker,
        mapCenter,
        setMapCenter,
        page,
        setPage,
      ]}
    >
      <div className="App">
        <div className="container">
          <TripsHeader setIsModalOpen={setIsModalOpen} />
          <SearchBar />
        </div>
        <div className="container">
          <TripsList />
          <Map />
        </div>
        {isModalOpen && (
          <Modal setIsModalOpen={setIsModalOpen} spaceCenters={spaceCenters} />
        )}
      </div>
    </AppContext.Provider>
  );
}

export default App;
