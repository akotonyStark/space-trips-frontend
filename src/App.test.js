import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import App from "./App";
import TripCard from "./components/TripCard";
import Map from "./components/Map";
import { AppContext } from "./App";

//app context values needed for testing some components

const marker = { id: "", isBouncing: false };
const hovered = { id: "", state: false };
const setHovered = (obj) => {
  return { ...obj, state: true };
};

const overrideStateArray = (state) => {
  let newState = [];
  state = newState;
  return state;
};
const overrideBooleanState = (state) => {
  let newState = true;
  state = newState;
  return state;
};
const overrideObjectState = (state) => {
  let newState = {};
  state = newState;
  return state;
};
const setResponse = (response) => {
  let newState = "some response";
  response = newState;
  return response;
};

const departureDate = new Date("2022-12-07");
const trips = [];
const mapCenter = [];
const viewState = {
  width: "100%",
  height: 900,
  latitude: 41.579606918652054,
  longitude: 4.244298260567439,
  zoom: 3.5,
  bearing: 0,
  pitch: 0,
  transitionDuration: 1000,
};

//component rendering test
test("renders the main App component", () => {
  render(<App />);
  const appTitle = screen.getByText(/SPACE TRIPS/i);
  expect(appTitle).toBeInTheDocument();
});

test("if Space Centers render and shows space center details", () => {
  render(
    <AppContext.Provider
      value={{
        setHovered,
        marker,
        setMapCenter: overrideStateArray,
        departureDate,
        setFlights: overrideStateArray,
        setShowFlightsList: overrideBooleanState,
        setshowSearching: overrideBooleanState,
        showNotification: overrideBooleanState,
        setResponse,
      }}
    >
      <TripCard
        spaceCenter={{
          id: 1,
          name: "Accra Space Center",
          planet: { name: "Earth" },
        }}
        marker={{ id: "", isBouncing: false }}
      />
    </AppContext.Provider>
  );
  expect(screen.getByText(/Accra Space Center/i)).toBeInTheDocument();
  expect(screen.getByText(/Earth/i)).toBeInTheDocument();
});

test("if map component renders", () => {
  render(
    <AppContext.Provider
      value={{
        trips,
        viewState,
        setViewState: overrideObjectState,
        hovered,
        setMarker: overrideObjectState,
        mapCenter,
      }}
    >
      <Map />
    </AppContext.Provider>
  );
});

//user interaction test
test("if hover effect works correctly", () => {
  const component = renderer.create(
    <AppContext.Provider
      value={{
        setHovered,
        marker,
        setMapCenter: overrideStateArray,
        departureDate,
        setFlights: overrideStateArray,
        setShowFlightsList: overrideBooleanState,
        setshowSearching: overrideBooleanState,
        showNotification: overrideBooleanState,
        setResponse,
      }}
    >
      <TripCard
        spaceCenter={{
          id: 1,
          name: "Accra Space Center",
          planet: { name: "Earth" },
        }}
        marker={{ id: "", isBouncing: false }}
        setHovered={setHovered}
      />
    </AppContext.Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  //console.log(tree.children[0].props);
  //manually trigger the callback
  tree.children[0].props.onMouseEnter();
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger the callback
  tree.children[0].props.onMouseLeave();
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
