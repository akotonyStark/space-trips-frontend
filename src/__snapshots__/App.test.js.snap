// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if hover effect works correctly 1`] = `
<div>
  <div
    className="sc-iCfMLu foJxYF trip-card"
    id="Accra-Space-Center"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      style={
        Object {
          "padding": 10,
        }
      }
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "justifyContent": "space-between",
          }
        }
      >
        <span
          className="trip-title"
        >
          Accra Space Center
        </span>
        <img
          alt="Rocket.svg"
          className="rocket"
          src="Rocket.svg"
        />
      </div>
      <span
        className="trip-planet"
      >
        Earth
      </span>
      <p
        style={
          Object {
            "marginTop": 30,
          }
        }
      >
        12 departures planned today
      </p>
    </div>
    <div
      className="sc-furwcr eqwTlf"
      onClick={[Function]}
      title="click to view all flights"
    >
      SEE ALL FLIGHTS
    </div>
  </div>
</div>
`;

exports[`if hover effect works correctly 2`] = `
<div>
  <div
    className="sc-iCfMLu foJxYF trip-card"
    id="Accra-Space-Center"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      style={
        Object {
          "padding": 10,
        }
      }
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "justifyContent": "space-between",
          }
        }
      >
        <span
          className="trip-title"
        >
          Accra Space Center
        </span>
        <img
          alt="Rocket.svg"
          className="rocket"
          src="Rocket.svg"
        />
      </div>
      <span
        className="trip-planet"
      >
        Earth
      </span>
      <p
        style={
          Object {
            "marginTop": 30,
          }
        }
      >
        12 departures planned today
      </p>
    </div>
    <div
      className="sc-furwcr eqwTlf"
      onClick={[Function]}
      title="click to view all flights"
    >
      SEE ALL FLIGHTS
    </div>
  </div>
</div>
`;

exports[`if hover effect works correctly 3`] = `
<div>
  <div
    className="sc-iCfMLu foJxYF trip-card"
    id="Accra-Space-Center"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      style={
        Object {
          "padding": 10,
        }
      }
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "justifyContent": "space-between",
          }
        }
      >
        <span
          className="trip-title"
        >
          Accra Space Center
        </span>
        <img
          alt="Rocket.svg"
          className="rocket"
          src="Rocket.svg"
        />
      </div>
      <span
        className="trip-planet"
      >
        Earth
      </span>
      <p
        style={
          Object {
            "marginTop": 30,
          }
        }
      >
        12 departures planned today
      </p>
    </div>
    <div
      className="sc-furwcr eqwTlf"
      onClick={[Function]}
      title="click to view all flights"
    >
      SEE ALL FLIGHTS
    </div>
  </div>
</div>
`;
